packages/xrpc-server/src/stream/subscription.ts

```diff
@@ -0,0 +1,137 @@
+import { wait } from '@atproto/common'
+import { WebSocket, ClientOptions } from 'ws'
+import { byMessage } from './stream'
+import { CloseCode } from './types'
+
+export class Subscription<T = unknown> {
+  constructor(
+    public opts: ClientOptions & {
+      service: string
+      method: string
+      maxReconnectSeconds?: number
+      validate: (obj: unknown) => T | undefined
+      getParams?: () =>
+        | Record<string, unknown>
+        | Promise<Record<string, unknown> | undefined>
+        | undefined
+    },
+  ) {}
+
+  async *[Symbol.asyncIterator](): AsyncGenerator<T> {
+    let reconnects: number | null = null
+    const maxReconnectMs = 1000 * (this.opts.maxReconnectSeconds ?? 64)
+    while (true) {
+      if (reconnects !== null) {
+        await wait(backoffMs(reconnects++, maxReconnectMs))
+      }
+      const ws = await this.getSocket()
+      ws.once('open', () => (reconnects = 0))
+      ws.once('close', (code, reason) => {
+        if (code === CloseCode.Abnormal) {
+          // Forward into an error to distinguish from a clean close
+          ws.emit(
+            'error',
+            new AbnormalCloseError(`Abnormal ws close: ${reason.toString()}`),
+          )
+        }
+      })
+      try {
+        for await (const message of byMessage(ws)) {
+          const result = this.opts.validate(message.body) // @TODO map $type
+          if (result !== undefined) {
+            yield result
+          }
+        }
+      } catch (err) {
+        ws.close() // No-ops if already closed or closing
+        if (isReconnectable(err, reconnects)) {
+          reconnects ??= 0 // Never reconnect with a null
+          continue
+        } else {
+          throw err
+        }
+      }
+      break // Other side cleanly ended stream and disconnected
+    }
+  }
+
+  private async getSocket() {
+    const params = (await this.opts.getParams?.()) ?? {}
+    const query = encodeQueryParams(params)
+    const url = `${this.opts.service}/xrpc/${this.opts.method}?${query}`
+    return new WebSocket(url, this.opts)
+  }
+}
+
+export default Subscription
+
+class AbnormalCloseError extends Error {
+  code = 'EWSABNORMALCLOSE'
+}
+
+function isReconnectable(err: unknown, reconnects: number | null): boolean {
+  // The socket never opened on initial connection.
```

Comment:
shouldn't we try to reconnect if initial connection failed due to network error?

Reply:
I struggled with this one.  The thing you do ideally want to be able to catch on init is any form of misconfiguration.  So then the question becomes, how do you distinguish misconfiguration from downtime on the producer side?  And that becomes hard to answer.  Any thoughts on how to slice this?

Reply:
I talked to @Jacob2161 about this a little bit.  He recommended having reconnection logic for the initial connection but maybe making it a bit more aggressive than the backoff strategy (e.g. repeatedly trying once a second) that occurs when you lose an existing connection.

If the app view weren't also serving XRPC requests I think it would make more sense to hard error and let Docker recycle the container as I'm doing right now, but since that's not the case I agree that we should have reconnection logic on the initial connection.  To mitigate the misconfiguration question I can make sure there's good logging going on.

Reply:
nice that all sounds great to me 👍